lib/segment/src/index/field_index/geo_index.rs

```diff
@@ -1 +1,531 @@
+const HASH_GRID_SIZE: &[(f64, f64)] = &[
+    (5009.4 * 1000.0, 4992.6 * 1000.0),
+    (1252.3 * 1000.0, 624.1 * 1000.0),
+    (156.5 * 1000.0, 156.0 * 1000.0),
+    (39.1 * 1000.0, 19.5 * 1000.0),
+    (4.9 * 1000.0, 4.9 * 1000.0),
+    (1.2 * 1000.0, 609.4),
+    (152.9, 152.4),
+    (38.2, 19.),
+    (4.8, 4.8),
+    (1.2, 0.595),
+    (0.149, 0.149),
+    (0.037, 0.019),
+];
+
+impl From<GeoPoint> for Coordinate<f64> {
+    fn from(point: GeoPoint) -> Self {
+        Self {
+            x: point.lat,
+            y: point.lon,
+        }
+    }
+}
+
+#[derive(Debug)]
+struct FullGeoBoundingBox {
+    north_west: GeoPoint,
+    south_west: GeoPoint,
+    south_east: GeoPoint,
+    north_east: GeoPoint,
+}
+
+impl FullGeoBoundingBox {
+    fn from_geo_bounding_box(bounding_box: &GeoBoundingBox) -> Self {
+        let GeoPoint {
+            lat: max_lat,
+            lon: min_lon,
+        } = bounding_box.top_left;
+        let GeoPoint {
+            lat: min_lat,
+            lon: max_lon,
+        } = bounding_box.bottom_right;
+
+        let north_west = GeoPoint {
+            lon: min_lon,
+            lat: max_lat,
+        };
+        let south_west = GeoPoint {
+            lon: min_lon,
+            lat: min_lat,
+        };
+        let south_east = GeoPoint {
+            lon: max_lon,
+            lat: min_lat,
+        };
+        let north_east = GeoPoint {
+            lon: max_lon,
+            lat: max_lat,
+        };
+
+        Self {
+            north_west,
+            south_west,
+            south_east,
+            north_east,
+        }
+    }
+
+    fn shortest_side_length_in_km(&self) -> f64 {
+        // the projection on the sphere is a trapeze - calculate 3 distances
+        let upper_width = distance_in_km_between_coordinates(&self.north_west, &self.north_east);
+        let lower_width = distance_in_km_between_coordinates(&self.south_west, &self.south_east);
+        // both height sides are equal in a trapeze - pick only one
+        let height = distance_in_km_between_coordinates(&self.north_west, &self.south_west);
+        f64::min(f64::min(upper_width, lower_width), height)
+    }
+}
+
+#[derive(Debug)]
+struct RectangleGeoHash {
+    north_west: GeoHash,
+    south_west: GeoHash,
+    south_east: GeoHash,
+    north_east: GeoHash,
+}
+
+impl RectangleGeoHash {
+    /// warning: this could generate a huge result depending on the size of the rectangle and the precision of the geohashes
+    fn geohash_regions(&self) -> Vec<GeoHash> {
+        let width = self.area_width_region_count();
+        let height = self.area_height_region_count();
+
+        let mut seen: Vec<String> = Vec::new();
+        let mut top = self.north_west.clone();
+        // traverse square by columns top to bottom
+        for _i in 0..width {
+            let mut current = top.clone();
+            for _j in 0..height {
```

Comment:
```suggestion
        for _ in 0..width {
            let mut current = top.clone();
            for _ in 0..height {
```
Minor stylistic suggestion